common/acl/camanager.go

```diff
@@ -0,0 +1,311 @@
+// copy and update from https://github.com/kubeedge/kubeedge/blob/c77318a1c3a97d63bcc3d5fd6cf4607d5df939ff/edge/pkg/edgehub/certificate/certmanager.go
+
+package acl
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/hex"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"io/ioutil"
+	nethttp "net/http"
+	"os"
+	"strings"
+
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"k8s.io/client-go/util/cert"
+	"k8s.io/klog/v2"
+
+	meshConstants "github.com/kubeedge/edgemesh/common/constants"
+	"github.com/kubeedge/kubeedge/common/constants"
+	"github.com/kubeedge/kubeedge/edge/pkg/edgehub/common/certutil"
+	"github.com/kubeedge/kubeedge/edge/pkg/edgehub/common/http"
+)
+
+const (
+	retry = 3
+)
+
+type caManager struct {
+	CR *x509.CertificateRequest
+
+	caFile   string
+	certFile string
+	keyFile  string
+
+	token string
+
+	caURL   string
+	certURL string
+}
+
+func init() {
+	constructMap[TypeWithCA] = newCAManager
+}
+
+// newCAManager creates a caManager for edge acl management according to EdgeHub config
+func newCAManager(tunnel TunnelACLConfig) Manager {
+	certReq := &x509.CertificateRequest{
+		Subject: pkix.Name{
+			Country:      []string{"CN"},
+			Organization: []string{"kubeEdge"},
+			Locality:     []string{"Hangzhou"},
+			Province:     []string{"Zhejiang"},
+			CommonName:   "kubeedge.io",
+		},
+	}
+	return &caManager{
+		token:    tunnel.Token,
+		CR:       certReq,
+		caFile:   tunnel.TLSCAFile,
+		certFile: tunnel.TLSCertFile,
+		keyFile:  tunnel.TLSPrivateKeyFile,
+		caURL:    tunnel.HTTPServer + constants.DefaultCAURL,
+		certURL:  tunnel.HTTPServer + constants.DefaultCertURL,
+	}
+}
+
+// Start starts the caManager
+func (m *caManager) Start() {
+	_, err := m.getCurrentCert()
+	if err != nil {
+		klog.Infof("acl cert, key file are no exist, start generate and fetch")
+		err = m.applyCerts()
+		if err != nil {
+			klog.Fatalf("Error: %v", err)
+		}
+	} else {
+		klog.Infof("acl cert, key file already exist, skip generate")
+	}
+}
+
+func (m *caManager) Name() string {
+	return "caManager"
+}
+
+// getCurrentCert returns current edge certificate
+func (m *caManager) getCurrentCert() (*tls.Certificate, error) {
+	cert, err := tls.LoadX509KeyPair(m.certFile, m.keyFile)
+	if err != nil {
+		return nil, err
+	}
+	if len(cert.Certificate) == 0 {
+		return nil, fmt.Errorf("cert.Certificate size is zero")
```

Comment:
Which cert is zero?

Reply:
Actually, if the certs is valid, no cert will be zero.